### PR TITLE
Fix cast of generated partitions

### PR DESCRIPTION
**Pull request by mfussenegger**
A cast can convert a `GeneratedReference` to a `SimpleReference` - that
broke a `indexOf(partitionColumns, column)` call when resolving the
reference implementation.

Closes https://github.com/crate/crate/issues/14307

---

First commit fixes the issue by making `LuceneReferenceResolver` more lenient in regards to concrete Reference instances, but I added another commit to also preserve the `GeneratedReference` instance on casts. This should help ensure that we don't have other places where a `.contains` or `.indexOf` broke.

### CODE REVIEW

**Review by seut — APPROVED** [review]
👍

**Comment by BaurzhanSakhariev** [comment]
failure seems related?

**Comment by mfussenegger** [comment]
Had to disable jdbc for the test, because it applied timezone logic to the timestamp and broke the comparison.

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 1]
```diff
@@ -2419,4 +2420,24 @@ public void test_set_number_of_replicas_on_partitioned_table() {
...
+        execute("""
+            CREATE TABLE tbl (
+                ts TIMESTAMP,
+                year TIMESTAMP GENERATED ALWAYS AS date_trunc('year',ts)
+            ) PARTITIONED BY (year)
+            """
+        );
+        execute("INSERT INTO tbl (ts) SELECT now()");
+        execute("refresh table tbl");
+        //execute("explain select year, year::TEXT from tbl LIMIT 10");
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2436]: leftover?

**Pull request merged**